bugwarrior/config/schema.py

```diff
@@ -16,133 +28,149 @@
 log = logging.getLogger(__name__)
 
 
-class StrippedTrailingSlashUrl(pydantic.v1.AnyUrl):
-    @classmethod
-    def validate(cls, value, field, config):
-        return super().validate(value.rstrip('/'), field, config)
+def validate_url(url: str):
+    return str(AnyUrl(url)).rstrip("/")
 
 
-class UrlSchemeError(pydantic.v1.UrlSchemeError):
-    msg_template = "URL should not include scheme ('{scheme}')"
+# NOTE: breaking change in pydantic v2: AnyUrl is not longer a subclass of str
+StrippedTrailingSlashUrl = Annotated[str, BeforeValidator(validate_url)]
 
 
-class NoSchemeUrl(StrippedTrailingSlashUrl):
-    @classmethod
-    def validate_parts(cls, parts: typing.Dict[str, str]) -> typing.Dict[str, str]:
-        scheme = parts['scheme']
-        if scheme is not None:
-            raise UrlSchemeError(scheme=scheme)
+def validate_no_scheme_url(value: str) -> str:
+    if "://" in value:
+        scheme = value.split("://")[0]
+        raise PydanticCustomError(
+            "url_scheme_not_allowed",
+            "URL should not include scheme ('{scheme}')",
+            {"scheme": scheme},
+        )
 
-        port = parts['port']
-        if port is not None and int(port) > 65_535:
-            raise pydantic.v1.errors.UrlPortError()
+    # Add a dummy scheme to leverage AnyUrl's validation (host, port, etc.)
+    # then validate as a proper URL
+    url_with_scheme = f"https://{value}"
 
-        user = parts['user']
-        if cls.user_required and user is None:
-            raise pydantic.v1.errors.UrlUserInfoError()
+    # NOTE: is this needed ?
+    try:
+        AnyUrl(url_with_scheme)
+    except Exception as exc:
+        raise PydanticCustomError(
+            "url_parsing", "Invalid URL format: {error}", {"error": str(exc)}
+        ) from exc
 
-        return parts
+    # NOTE: I did not migrate the user_required part, because it does not appear anywhere in the code
+    return value.rstrip("/")
 
 
-# Pydantic complicates the use of sets or lists as default values.
-class ConfigList(frozenset):
-    @classmethod
-    def __get_validators__(cls):
-        yield cls.validate
+NoSchemeUrl = Annotated[str, BeforeValidator(validate_no_scheme_url)]
 
-    @classmethod
-    def validate(cls, value):
-        """Cast ini string to a list of strings"""
+
+def parse_config_list(value: str | list[str]) -> list[str]:
+    """Cast ini string to a list of strings."""
+    if isinstance(value, str):
+        return [
+            item.strip()
+            for item in re.split(r",(?![^{]*})", value.strip())
+            if item != ""
+        ]
+    return list(value)
```

Comment:
It's already a list. Surely type checkers understand that?

```suggestion
    return value
```